### ISSUE TITLE
@gif in JupyterLab

### ISSUE THREAD

**Issue by OliverEvans96** [JuliaPlots/Plots.jl#1447]
Currently, `@gif` works fine in the standard notebook, but fails to display in JupyterLab.

See [here](https://github.com/JuliaPlots/Plots.jl/blob/master/src/animation.jl#L96):
```
# write out html to view the gif... note the rand call which is a hack so the image doesn't get cached
function Base.show(io::IO, ::MIME"text/html", agif::AnimatedGif)
    ext = file_extension(agif.filename)
    write(io, if ext == "gif"
        "<img src=\"$(relpath(agif.filename))?$(rand())>\" />"
    elseif ext in ("mov", "mp4")
        "<video controls><source src=\"$(relpath(agif.filename))?$(rand())>\" type=\"video/$ext\"></video>"
    else
        error("Cannot show animation with extension $ext: $agif")
    end)
end
```

It seems to be the `rand() ` trick to prevent caching that's causing the issue.
If `"<img src=\"$(relpath(agif.filename))?$(rand())>\" />"` is changed to `"<img src=\"$(relpath(agif.filename))\" />"`, then the gif displays fine.

I'd love to learn why this is acting differently in JupyterLab. Maybe @gnestor or @ian-r-rose can shed some light on the situation?

Thanks!
Oliver

**Comment by OliverEvans96** [comment]
Also, pinging @tbreloff since you wrote the hack :)

**Comment by ian-r-rose** [comment]
@OliverEvans96 If you use the browser inspector to look at the HTML for the produced output, does anything look wrong? Sometimes the HTML sanitizer can do unexpected things...

**Comment by OliverEvans96** [comment]
Ah, good catch - it is an HTML sanitization issue. Without removing the `rand()`, the `img` tag has `src="http://localhost:8888/files/academic/research/droplet/odes/tmp.gif%3F0.9281699715131861%3E"`, whereas in the classic notebook, the same output refers to `src="tmp.gif?0.9281699715131861>"`.

If I copy and paste `http://localhost:8888/files/academic/research/droplet/odes/tmp.gif%3F0.9281699715131861%3E` into my browser, it gives a 404, but changing `%3F0` to `?` shows the gif.

Is there a recommended fix for this type of thing?

**Comment by ian-r-rose** [comment]
Ah, I now don't think it is a sanitization issue. I think it's related to jupyterlab/jupyterlab#3851.

**Comment by OliverEvans96** [comment]
I'm not sure I understand. It looks like in the commit you're referring to, you're already handling this type of cache busting by default in lab. But [this line (737)](https://github.com/jupyterlab/jupyterlab/pull/3851/files#diff-893eb72717d19ccc10c070e85bc7cd04R737) makes me think that busting the cache twice shouldn't cause an issue.

It seems like if there were a way to prevent `?` from being converted to `%3F0`, the issue wouls be solved, correct?

**Comment by OliverEvans96** [comment]
I agree - this seems like a frontend issue. @tbreloff, what do you think of merging #1448 and supporting cache-busting in Jupyter in jupyter/notebook#1369 ?

**Comment by gnestor** [comment]
In JupyterLab, I'm seeing a query string automatically appended to images displayed using ipython:

```py
from IPython.display import Image

Image(url='/giphy.gif')
```

results in 

```html
<img src="http://localhost:8888/files/giphy.gif?1547629927391" alt="Image">
```

I'm not sure if this is ipython appending this query string on jupyterlab.

**Comment by darthdeus** [comment]
@gnestor This doesn't seem to happen in regular Jupyter. Is there any chance for the bugfix being backported from Jupyer Lab?

**Comment by gnestor** [comment]
We could certainly backport this although there are very few new features being added to classic notebook. This could also be done in ipython too (and perhaps it should be). To do this in Python yourself:

```py
from IPython.display import Image
import time

Image(url=f'../giphy.gif?{str(int(time.time()))}')
```
